Fix JP Screenshot Automation UI Tests

ovitrif opened the pull request:
Fixes #16845 together with the changes from #16846.

Since it builds on top of that PR, this is targeting the branch of that one.

Changes:
- Fix mocks for jetpack screenshot tests
- Fix typo in new features dialog layout
- Fix login in Jetpack UI tests (and dismiss full screen dialog about insights)
- Fix navigation to activity log in JP screenshots test
- Fix navigation to stats in JP screenshots test
- Use recycler view to click stats and media in site menu
- Use only one mock for `/me/sites` (WP&JP) & adapt mocks to correct site id
- Add `switchToSite` in `MySitePage` and use it
- Switch to the correct site in `navigateDiscover` (since `editBlogPost` is commented)
- Refactor `goToPosts` to use recycler view item click
- Make navigation to site menu items consistent
- Use `goToStats` instead of `clickStats`, and removed `clickStats`.
- Improve code to navigate back from media in WP screenshot tests
- Various fixes to make navigation to site menu items work when Quick Actions are displayed or not
- Dismiss dialog about Insights on Stats screen

To test:
1. Checkout the PR branch locally ([Github CLI](https://cli.github.com/) command: `gh pr checkout 16877`)
2. Switch to `jetpackWasabiDebug` build flavor.
3. Run the `jPScreenshotTest` from `JPScreenshotTest` in Android Studio to check that it **succeeds**.

   ⚠️ **Note:** _When running multiple times it fails too often because the mock for activity is not always applying, which is pretty strange but I didn't have time to look deeper into the reason behind this.
This can fail in more places than `navigateActivityLog`, sometimes it loads correctly for activity log but then won't load for one of the next tests._

   <details>
   <summary>Emulator Setup Guide</summary>

     Ideally we run the tests on an emulator created by fastlane. See the notes in the issue for how this can be done:

     > Follow the guide in `SCREENSHOT_DEVICE_SETUP.md` to create a device like the ones we're using for screenshots automation (this won't be super easy as the guide is a bit outdated and if you're on an M1 Mac you have to change the generated .ini files manually to set the arm64 architecture or else the emultators won't start.
   </details>

4. Close emulator & wipe data
5. Switch to `wordpressWasabiDebug` build flavor.
6. Run the `wPScreenshotTest` from `WPScreenshotTest` to check that it **succeeds**, meaning we're regression free.

## Regression Notes
1. Potential unintended areas of impact
   - WP Screenshot tests.
   - `EditorTests`
   - `allDayStatsLoad()`

2. What I did to test those areas of impact (or what existing automated tests I relied on)
   - Ran WP Screenshot tests and fixed the issues by using only one mock for `/me/sites`.
   - `EditorTests` are disabled.
   - `allDayStatsLoad` fails, but it also fails without the changes.

3. What automated tests I added (or what prevented me from doing so)
   - It's only automated UI tests changes 😅.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ovitrif committed:
Fix mocks for jetpack screenshot tests

ovitrif committed:
Fix typo in new features dialog

ovitrif committed:
Fix login in Jetpack UI tests

ovitrif committed:
Fix navigation to activity log in JP screenshots test

ovitrif committed:
Fix navigation to stats in JP screenshots test

ovitrif committed:
Use recycler view to click stats and media in site menu

ovitrif committed:
Use only one mock for /me/sites (WP&JP) + adapt mocks to correct site id

ovitrif committed:
Add switchToSite in MySitePage and use it

ovitrif committed:
Switch to the correct site in navigateDiscover (since editBlogPost is commented)

ovitrif committed:
Refactor goToPosts to use recycler view item click

ovitrif committed:
Make navigation to site menu items consistent

ovitrif committed:
Use goToStats instead of clickStats

ovitrif committed:
Fix navigation back from media

ovitrif committed:
Remove unused import to fix checkstyle lint

ovitrif commented:
Looking at the 300 LOC limit warning I get the feeling not all new mocks are needed anymore.

Some of them were added while still relying on the now removed `jetpack_rest_v11_me_sites.json` file.

I'll have a deeper look once I'm back on Monday.

zwarm commented on WordPress/src/main/res/layout/feature_introduction_dialog_fragment.xml:
```diff
@@ -76,7 +76,7 @@
             android:paddingEnd="@dimen/margin_extra_large"
             app:layout_behavior="@string/appbar_scrolling_view_behavior"
-            app:layout_constraintBottom_toTopOf="@+id/blogging_prompts_onboading_button_container"
+            app:layout_constraintBottom_toTopOf="@+id/blogging_prompts_onboarding_button_container"
```
@ovitrif - Nice catch 👍

zwarm commented on WordPress/src/androidTest/java/org/wordpress/android/e2e/flows/LoginFlow.java:
```diff
@@ -56,27 +57,22 @@ public void confirmLogin(boolean isSelfHosted) {
-
-            waitForElementToBeDisplayed(continueButton);
-            clickOn(continueButton);
+        // New Epilogue Screen - Choose the first site from the list of site.
```
@ovitrif Previously the experience was different for JP vs. WP. This works great now.

ovitrif replied:
Indeed, thanks for confirming this 🙇

zwarm commented on WordPress/src/androidTest/java/org/wordpress/android/e2e/pages/MySitesPage.java:
```diff
@@ -111,6 +111,7 @@ public MySitesPage switchToSite(String siteUrl) {
         return this;
     }
 
+    // Either this or goToStats should be removed after making the implementation uniform.
```
❓   Is there more coming for this PR to handle this comment; as I see both methods still hanging around? And yes, I did notice that `clickStats` is also being called from within `StatsTest`.

ovitrif replied:
I can address the issue in this PR once I'm resuming work on Monday 🙇🏻.
At the time of opening the PR and afterwards It was a bit difficult to find time for it.

ovitrif replied:
Monday update: Working on this 🚀

ovitrif replied:
Done. While on it I also had to fix (again 😅 ) the navigation to the site menu items since it was failing now that for an unclear reason the Quick Actions are visible again.

I adapted the navigation methods so they work both when the Quick Actions card is present and when it's not.

@zwarm Can you please check the changes again and if all is ok now we're going to merge this into the target branch 🙇

ovitrif replied:
I had to do more fixes and I also cleaned up the mocks by removing the ones referring to non existing site IDs.

ovitrif replied:
> I am still seeing the grey background, but I know that this will be addressed separately (after this PR is merged into the fix/screenshots-test branch).

@zwarm Will look into it once we're done merging this into `fix/screenshots-test` 🙇 .

zwarm reviewed (approved):
@ovitrif 
- All works as expected - the tests run to completion; however  like you pointed out, if I run multiple times in a row, the test fails. The only solution to get them working again, for me,  was to rebuild the app and install on the device BEFORE re-running the tests. My error didn't appear to be caused by wiremock, but instead `No views in hierarchy found matching`. Perhaps it is wiremock. During my foray into writing the original JP screenshot tests, I ran into quite a few quirky issues. The worst were while running via fastlane and for that I had to run `adb kill-server` followed by `adb start-server` to get things to work again. I'm not 100% certain that we'll be able to get to 0 failures and maybe that is why these tests were never added as connected tests. 🤔 

- Nice work switching over to use the recycler view for site menu items. Glad it worked out.
- I am still seeing the grey background, but I know that this will be addressed separately (after this PR is merged into the `fix/screenshots-test` branch).

I am approving this PR, but will not merge it because I'm not sure if you  wanted to address `goToStats` vs. `clickStats` functionality here. If not, feel free to merge.

ovitrif commented:
> My error didn't appear to be caused by wiremock, but instead No views in hierarchy found matching. Perhaps it is wiremock.

I think it could be. In my case the error was on one of the screens where the activity log is supposed to be loaded, but it was empty, thus the expected elements that are to be clicked by the tests are missing, which gives us the error with views not being present at the time where needed.

>  I'm not 100% certain that we'll be able to get to 0 failures and maybe that is why these tests were never added as connected tests. 🤔

Noted, we could do an effort as long as it's timeboxed, but I get the same feeling as you've shared here that it might be not trivial and thus perhaps not worth the effort in terms of ROI to invest too much time into getting the tests stable enough to have them run as connected tests.

ovitrif committed:
Merge branch 'trunk' into fix/screenshot-tests-jetpack

ovitrif committed:
Refactor goToStats to remove clickStats duplicate

ovitrif committed:
Refactor navigation to site menu items to work when quick actions are available

ovitrif committed:
Remove unused import in MySitesPage

ovitrif committed:
Merge branch 'fix/screenshots-tests' into fix/screenshot-tests-jetpack

ovitrif committed:
Remove all mocks of removed site with id 185119032

ovitrif committed:
Replace all refs of site id 185124945 to site id 181977606 in mocks

ovitrif committed:
Call switchToSite in navigateMySite from JpScreenshotTests

ovitrif committed:
Remove all mocks of removed site with id 185119569

ovitrif committed:
Fix navigation to backup when quick actions are shown

ovitrif committed:
Fix typo in first param of clickQuickActionOrSiteMenuItem

ovitrif committed:
Dismiss UpdateAlertDialogFragment if displayed on stats screen

ovitrif committed:
Remove unused imports from StatsPage

ovitrif commented:
> Looking at the 300 LOC limit warning I get the feeling not all new mocks are needed anymore.
> 
> Some of them were added while still relying on the now removed `jetpack_rest_v11_me_sites.json` file.
> 
> I'll have a deeper look once I'm back on Monday.

Was able to get rid of some mocks and I also updated a few references to site ID's that we've removed with this PR.

zwarm commented on WordPress/src/androidTest/java/org/wordpress/android/e2e/pages/MySitesPage.java:
```diff
@@ -139,4 +141,18 @@ public static void goToHomeTab() {
+     * @param quicActionItemId Id of the quick actions menu item.
+     * @param siteMenuItemString String resource id of the site menu item.
+     */
+    private void clickQuickActionOrSiteMenuItem(@IdRes int quicActionItemId, @StringRes int siteMenuItemString) {
```
@ovitrif - great idea on this method. I think there _could_ be an experiment on quick actions + it might work different for WP vs. JP. 🤔  experiments and tests are not a good combination.

ovitrif replied:
Thank you 🙇 

> experiments and tests are not a good combination

Got to the same conclusion 😅

zwarm reviewed (approved):
👋 @ovitrif - I have reviewed the last set of changes and have run the tests. No new issues uncovered. I agree with you on the wiremock issues re: activityLog ... which btw I have notice, is sometimes slow to load outside of running tests. 

I like the improvements to use StringRes that you added to the Utils class. Plus all the adjustments to the mocks. 👍 

I think this one is good to go. Thanks again for tackling this. 
:)
